Update celery routing for celery 4+

iamsobanjaved opened the pull request:
This is the new PR for celery routing as that revert PR's branch name had too many reverts in it. ~~We are currently facing an `UnregisteredTask` exception on the sandbox after rebasing with the latest master. Although this exception disappears after restarting sandbox by `sudo /edx/bin/supervisorctl restart all` multiple times~~

Now that fix is merged we can merge this.

This reverts commit c1fe3c3a93eecaca1fcd70af2c4c769cf6d43eb5.
(cherry picked from commit db4c3b1210d8826df417e1d2c6ced72dd85fea8b)
https://github.com/edx/edx-platform/pull/25565

awais786 commented:
Some thing wrong with the PR sandbox throwing error.
```
The full contents of the message body was:
b'{"task": "lms.djangoapps.instructor_task.tasks.calculate_grades_csv", "id": "b35980b0-9724-49c7-947e-d28761b8bcc8", "args": [1, {"xqueue_callback_url_prefix": "https://iamsobanjaved.sandbox.edx.org", "request_info": {"username": "staff", "user_id": 13, "ip": "111.119.187.19", "agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.183 Safari/537.36", "host": "iamsobanjaved.sandbox.edx.org"}, "task_id": "b35980b0-9724-49c7-947e-d28761b8bcc8"}], "kwargs": {}, "group": null, "group_index": null, "retries": 0, "eta": null, "expires": null, "utc": true, "callbacks": null, "errbacks": null, "timelimit": [null, null], "taskset": null, "chord": null}' (708b)
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/celery/worker/consumer/consumer.py", line 562, in on_task_received
    strategy = strategies[type_]
KeyError: 'lms.djangoapps.instructor_task.tasks.calculate_grades_csv'
````

nedbat commented:
@jmbowman This is the latest PR for the routing changes.  Is it bad to make Koa without it?

iamsobanjaved commented:
We have made sandbox with master and it was working fine, but after some restarts, it has started to showing `UnregisteredTask` exception on Instructor tasks. Even though it is showing registered when we run this `DJANGO_SETTINGS_MODULE=lms.envs.production /edx/app/edxapp/venvs/edxapp/bin/python3.8 /edx/app/edxapp/venvs/edxapp/bin/celery --app=lms.celery:APP  inspect registered`

Output from this command:
`-> celery@edx.lms.core.high_mem.mastertest: OK
    * celery_utils.tasks.mark_resolved
    * edx_sga.tasks.zip_student_submissions
    * enterprise.tasks.create_enterprise_enrollment
    * entitlements.expire_old_entitlements [routing_key=edx.lms.core.default]
    * integrated_channels.integrated_channel.tasks.transmit_content_metadata
    * integrated_channels.integrated_channel.tasks.transmit_learner_data
    * integrated_channels.integrated_channel.tasks.transmit_single_learner_data
    * integrated_channels.integrated_channel.tasks.unlink_inactive_learners
    * lms.djangoapps.bulk_email.tasks.send_course_email
    * lms.djangoapps.ccx.tasks.send_ccx_course_published
    * lms.djangoapps.certificates.tasks.generate_certificate
    * lms.djangoapps.discussion.tasks.send_ace_message [routing_key=edx.lms.core.default]
    * lms.djangoapps.discussion.tasks.update_discussions_map
    * lms.djangoapps.email_marketing.tasks.get_email_cookies_via_sailthru [routing_key=edx.lms.core.default]
    * lms.djangoapps.email_marketing.tasks.update_course_enrollment [routing_key=edx.lms.core.default]
    * lms.djangoapps.email_marketing.tasks.update_user [routing_key=edx.lms.core.default]
    * lms.djangoapps.email_marketing.tasks.update_user_email [routing_key=edx.lms.core.default]
    * lms.djangoapps.gating.tasks.task_evaluate_subsection_completion_milestones
    * lms.djangoapps.grades.tasks.compute_all_grades_for_course [routing_key=edx.lms.core.default]
    * lms.djangoapps.grades.tasks.compute_grades_for_course
    * lms.djangoapps.grades.tasks.compute_grades_for_course_v2 [rate_limit=300/h]
    * lms.djangoapps.grades.tasks.recalculate_course_and_subsection_grades_for_user [routing_key=edx.lms.core.default]
    * lms.djangoapps.grades.tasks.recalculate_subsection_grade_v3 [routing_key=edx.lms.core.default]
    * lms.djangoapps.instructor_task.tasks.calculate_grades_csv [routing_key=edx.lms.core.high_mem]
    * lms.djangoapps.instructor_task.tasks.calculate_may_enroll_csv
    * lms.djangoapps.instructor_task.tasks.calculate_problem_grade_report [routing_key=edx.lms.core.high_mem]
    * lms.djangoapps.instructor_task.tasks.calculate_problem_responses_csv.v2 [routing_key=edx.lms.core.high_mem]
    * lms.djangoapps.instructor_task.tasks.calculate_students_features_csv
    * lms.djangoapps.instructor_task.tasks.cohort_students
    * lms.djangoapps.instructor_task.tasks.course_survey_report_csv
    * lms.djangoapps.instructor_task.tasks.delete_problem_state
    * lms.djangoapps.instructor_task.tasks.export_ora2_data
    * lms.djangoapps.instructor_task.tasks.export_ora2_submission_files
    * lms.djangoapps.instructor_task.tasks.generate_certificates [routing_key=edx.lms.core.high_mem]
    * lms.djangoapps.instructor_task.tasks.override_problem_score
    * lms.djangoapps.instructor_task.tasks.proctored_exam_results_csv
    * lms.djangoapps.instructor_task.tasks.rescore_problem
    * lms.djangoapps.instructor_task.tasks.reset_problem_attempts
    * lms.djangoapps.instructor_task.tasks.send_bulk_course_email
    * lms.djangoapps.program_enrollments.tasks.expire_waiting_enrollments
    * lms.djangoapps.verify_student.tasks.send_request_to_ss_for_user [routing_key=edx.lms.core.high]
    * lms.djangoapps.verify_student.tasks.send_verification_status_email [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.content.block_structure.tasks.get_course_in_cache
    * openedx.core.djangoapps.content.block_structure.tasks.get_course_in_cache_v2
    * openedx.core.djangoapps.content.block_structure.tasks.update_course_in_cache
    * openedx.core.djangoapps.content.block_structure.tasks.update_course_in_cache_v2
    * openedx.core.djangoapps.coursegraph.tasks.dump_course_to_neo4j [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.credentials.tasks.v1.tasks.send_grade_to_credentials [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.programs.tasks.award_course_certificate [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.programs.tasks.award_program_certificates [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.programs.tasks.revoke_program_certificates [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.programs.tasks.update_certificate_visible_date_on_course_update [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks.ScheduleCourseNextSectionUpdate [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks.ScheduleCourseUpdate [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks.ScheduleRecurringNudge [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks.ScheduleUpgradeReminder [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks._course_update_schedule_send [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks._recurring_nudge_schedule_send [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks._upgrade_reminder_schedule_send [routing_key=edx.lms.core.default]
    * openedx.core.djangoapps.schedules.tasks.update_course_schedules
    * openedx.core.djangoapps.verified_track_content.tasks.sync_cohort_with_mode
    * openedx.features.enterprise_support.tasks.clear_enterprise_customer_data_consent_share_cache
    * student.send_activation_email
    * super_csv.mixins.do_deferred_commit
    * super_csv.tasks.expire_data
    * third_party_auth.fetch_saml_metadata
`

**Sandbox:** https://mastertest.sandbox.edx.org/
**Flower:** http://54.242.239.66:5555/dashboard

awais786 commented:
For a sandbox you can try these [commands](https://gist.github.com/awais786/2e835a9a8962193a28d00a2650bd6f73) it will execute the instructor tasks 'generate-grade-csv' and show the result.

jmbowman commented:
@nedbat I'm pretty sure we want this in Koa, but at this point we should just cherrypick the commit once we get the whole problem sorted out.  It shouldn't block any testing unless somebody wants to verify that tasks are still going to the correct queues (since without this they all go to the default queue).

@iamsobanjaved Yeah, the terrible part of this is that the problem is intermittent even within the same deployment, and celery gives no useful information whatsoever to help in debugging it.  I think we need to get some results from [BOM-2086](https://openedx.atlassian.net/browse/BOM-2086) before we can safely merge either this or the virtualenv upgrade again.

jmbowman commented:
Now that https://github.com/edx/edx-platform/pull/25840 has been merged, we should be clear to rebase and merge this.  That PR made some changes to these files, so you may need to refactor a bit.

iamsobanjaved commented:
After the fix, rebased it with the latest master and refactored it. Tested it on multiple sandboxes and `routing` is working as per expectations.

**Sandbox:** https://celeryroute2.sandbox.edx.org/
**Flower:** http://52.91.58.200:5555/dashboard

feanil reviewed (approved):
This looks great.  It looks as a result of this change, how people write celery tasks for LMS and CMS will change.  I think a follow-on task to write a new platform level `docs/howto`(How to add new celery tasks) would be a valuable doc to add.

Please make sure that that work is ticketed and done soon.

jmbowman commented:
@nedbat Do we want to cherry pick this into Koa to fix celery task routing?
